src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,972 @@
+                          aNeuronInnovations[innovIdx].newOutputLinkInnovId,
+                          genome.aLinkGenes[linkIdx].Weight(),
+                          true);
+      genome.AddLink(inputLink);
+      genome.AddLink(outputLink);
+      return;
+    }
+
+    // If new innovation, create.
+    NeuronInnovation neuronInnov = AddNeuronInnovation(splitLinkInnovId);
+
+    LinkInnovation inputLinkInnov;
+    inputLinkInnov.fromNeuronId = genome.aLinkGenes[linkIdx].FromNeuronId();
+    inputLinkInnov.toNeuronId = neuronInnov.newNeuronId;
+    inputLinkInnov.newLinkInnovId = neuronInnov.newInputLinkInnovId;
+    aLinkInnovations.push_back(inputLinkInnov);
+
+    LinkInnovation outputLinkInnov;
+    outputLinkInnov.fromNeuronId = neuronInnov.newNeuronId;
+    outputLinkInnov.toNeuronId = genome.aLinkGenes[linkIdx].ToNeuronId();
+    outputLinkInnov.newLinkInnovId = neuronInnov.newOutputLinkInnovId;
+    aLinkInnovations.push_back(outputLinkInnov);
+
+    NeuronGene neuronGene(neuronInnov.newNeuronId,
+                          HIDDEN,
+                          SIGMOID,  // TODO: make it random??
+                          (fromNeuron.Depth() + toNeuron.Depth()) / 2,
+                          0,
+                          0);
+    genome.AddHiddenNeuron(neuronGene);
+
+    LinkGene inputLink(genome.aLinkGenes[linkIdx].FromNeuronId(),
+                       neuronInnov.newNeuronId,
+                       neuronInnov.newInputLinkInnovId,
+                       1,
+                       true);
+    LinkGene outputLink(neuronInnov.newNeuronId,
+                        genome.aLinkGenes[linkIdx].ToNeuronId(),
+                        neuronInnov.newOutputLinkInnovId,
+                        genome.aLinkGenes[linkIdx].Weight(),
+                        true);
+    genome.AddLink(inputLink);
+    genome.AddLink(outputLink);
+  }
+
+  // Mutate: enable disabled, or disable enabled link.
+  void MutateEnableDisable(Genome& genome, bool enabled, double mutateProb) {
+    std::vector<ssize_t> linkIndexs;
+    for (ssize_t i=0; i<genome.NumLink(); ++i) {
+      if (genome.aLinkGenes[i].Enabled() == enabled) {
+        linkIndexs.push_back(i);
+      }
+    }
+
+    double p = mlpack::math::Random();
+    if (p<mutateProb && linkIndexs.size()>0) {
+      ssize_t idx = linkIndexs[mlpack::math::RandInt(0, linkIndexs.size())];
+      genome.aLinkGenes[idx].Enabled(!enabled);  // Reverse enabled status to opposite.
+    }
+  }
+
+  // Mutate: change single weight. Combine both biased and unbiased mutation.
+  void MutateWeight(Genome& genome, double mutateProb, double perturbProb, double mutateSize) {
+    double p = mlpack::math::Random();  // rand 0~1
+    if (p > mutateProb) return;
+    
+    for (ssize_t i=0; i<genome.aLinkGenes.size(); ++i) {  
+      double p2 = mlpack::math::Random();
+      if (p2 < perturbProb) {  // Biased weight mutation.
+        double deltaW = mlpack::math::RandNormal(0, mutateSize);
```

Comment:
I think RandomNormal is able to alter in both diractions, as it generates random Gaussian noise with mean 0. Isn't it?